lib/mayaUsd/ufe/UsdUndoableCommand.h

```diff
@@ -0,0 +1,197 @@
+//
+// Copyright 2020 Autodesk
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#pragma once
+
+#include <mayaUsd/base/api.h>
+#include <mayaUsd/undo/UsdUndoBlock.h>
+#include <mayaUsd/undo/UsdUndoableItem.h>
+
+#include <pxr/base/vt/value.h>
+#include <pxr/usd/usdGeom/xformOp.h>
+
+#include <ufe/baseUndoableCommands.h>
+#include <ufe/path.h>
+
+PXR_NAMESPACE_USING_DIRECTIVE
+
+namespace MAYAUSD_NS_DEF {
+namespace ufe {
+
+// Templated helper class to factor out common code for setting USD values.
+//
+// Implement the execute, undo and redo of the UFE command interfaces, with
+// common code proetecting against early undo/redo preceeding the initial
+// execution and declaring the UsdUndoBlock during the execution.
+//
+// Sub-classes should *not* have to override execute(), undo() nor redo().
+//
+// Sub-classes only need to implement the handleSet() function.
+//
+// Sub-classes can override set() if the command support interactive
+// manipulations. Since each set() in specific to the specific Cmd base
+// class, we cannot provide a generic implementation in this template.
+//
+// A typical set() implementation should call setNewValue() with the
+// new value and then execute() to actually set the value on the USD
+// attribute.
+template <typename Cmd> class UsdUndoableCmdBase : public Cmd
+{
+public:
+    UsdUndoableCmdBase(
+        const VtValue&     newOpValue,
+        const Ufe::Path&   path,
+        const UsdTimeCode& writeTime_);
+
+    // Ufe::UndoableCommand overrides. Do not override further.
+    void execute() override;
+    void undo() override;
+    void redo() override;
+
+    UsdTimeCode readTime() const;
+    UsdTimeCode writeTime() const;
+
+protected:
+    // Update the new value that will be set by execute().
+    void setNewValue(const VtValue& v);
+
```

Comment:
While the previous design of teh code used State inner classes, and initially it was thought that not all sub-classes would need it, it turns out UFE does not always call the function you would expect and sometimes makes extra call you would also not expect.

So the states were simplified to 3, and mostly it only switches between Done and Undone, really. So between all classes needing to be isolated from the UFE details and the small number of states, it seemed simplir to only have a enum state with a very few if in undo/redo/execute.